test/TestInternalGrains/StreamingImmutabilityTestGrain.cs

```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Threading.Tasks;
+using Orleans;
+using Orleans.Streams;
+using UnitTests.GrainInterfaces;
+
+namespace UnitTests.Grains
+{
+    public class StreamingImmutabilityTestGrain : Grain, IStreamingImmutabilityTestGrain
+    {
+        private StreamImmutabilityTestObject _myObject;
+        private StreamSubscriptionHandle<StreamImmutabilityTestObject> _streamSubscriptionHandle;
+
+        public async Task SubscribeToStream(Guid guid, string providerName)
+        {
+            var stream = GetStreamProvider(providerName).GetStream<StreamImmutabilityTestObject>(guid, "Namespace");
+            _streamSubscriptionHandle = await stream.SubscribeAsync(OnNextAsync);
+        }
+
+        public async Task UnsubscribeFromStream()
+        {
+            if (_streamSubscriptionHandle != null)
+                await _streamSubscriptionHandle.UnsubscribeAsync();
+        }
+
+        public async Task ProduceTestObject(string providerName)
+        {
+            _myObject = new StreamImmutabilityTestObject() {MyString = this.GetPrimaryKey().ToString()};
+            var stream = GetStreamProvider(providerName).GetStream<StreamImmutabilityTestObject>(this.GetPrimaryKey(), "Namespace");
+            await stream.OnNextAsync(_myObject);
```

Comment:
You're right, immutability includes two aspects in Orleans, of which I just tested aspect 1:
1. Immutable objects are transferred as is within the same silo and serialized between different silos.
2. As you mentioned application level promises not to modify the data.

I went with testing only aspect 1 because I wanted to make sure that disabling of immutability leads to correct serialization of `StreamImmutabilityTestObject`, so `MyString` is not transferred.

If there is need I can go and test aspect 2, but it should result from aspect 1 intentionally.